### PR TITLE
expression: fix issue that `date_add` and `date_sub` is incompatible with MySQL

### LINKED ISSUE #9667: Function `date_add` and `date_sub` don't check datetime field overflow
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
mysql> set sql_mode='traditional';
mysql> create table t1 (d date);
mysql> insert into t1 (d) select date_add('2000-01-01',interval 8000 year);
mysql> insert into t1 (d) select date_sub('2000-01-01', INTERVAL 2001 YEAR);
```

2. What did you expect to see?
In Mysql
```
mysql> set sql_mode='traditional';
Query OK, 0 rows affected (0.00 sec)
mysql> insert into t1 (d) select date_add('2000-01-01',interval 8000 year);
ERROR 1441 (22008): Datetime function: datetime field overflow
mysql> insert into t1 (d) select date_sub('2000-01-01', INTERVAL 2001 YEAR);
ERROR 1441 (22008): Datetime function: datetime field overflow
```

3. What did you see instead?
In TiDB
```
mysql> set sql_mode='traditional';
Query OK, 0 rows affected (0.00 sec)
mysql> insert into t1 (d) select date_add('2000-01-01',interval 8000 year);
Query OK, 1 row affected (0.00 sec)
Records: 1  Duplicates: 0  Warnings: 0
mysql> insert into t1 (d) select date_sub('2000-01-01', INTERVAL 2001 YEAR);
Query OK, 1 row affected (0.00 sec)
Records: 1  Duplicates: 0  Warnings: 0
```


4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                     |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.0-beta-197-g52c8267a7
Git Commit Hash: 52c8267a735dbb5e14b07c361ab8caa3efff5631
Git Branch: master
UTC Build Time: 2019-03-12 06:53:10
GoVersion: go version go1.12 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
fix #9667 

### What is changed and how it works?
Check if it `datetime` if overflow when `date_add` and `date_sub`.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: types/time.go [hunk 1]
```diff
@@ -2577,3 +2578,40 @@ func DateFSP(date string) (fsp int) {
 	}
 	return
 }
+
+var (
+	dateRange = []gotime.Time{
+		gotime.Date(1000, 1, 1, 0, 0, 0, 0, gotime.Local),
```
Comment (lamxTyler) [inline types/time.go]: Are they the same with https://github.com/pingcap/tidb/blob/master/types/time.go#L112?

File: types/time.go [hunk 2]
```diff
@@ -2577,3 +2578,40 @@ func DateFSP(date string) (fsp int) {
...
+
+// DateTimeIsOverflow return if this date is overflow.
+// See: https://dev.mysql.com/doc/refman/8.0/en/datetime.html
+func DateTimeIsOverflow(date Time) (bool, error) {
+	var b, e gotime.Time
+	switch date.Type {
+	case mysql.TypeDate:
+		b, e = dateRange[0], dateRange[1]
+	case mysql.TypeTimestamp:
+		b, e = timestampRange[0], timestampRange[1]
```
Comment (zz-jason) [inline types/time.go]: should we consider the system timezone info for `timestamp` values?

File: types/time.go [hunk 3]
```diff
@@ -2577,3 +2578,40 @@ func DateFSP(date string) (fsp int) {
...
 }
+
+var (
+	dateRange = []gotime.Time{
+		gotime.Date(1000, 1, 1, 0, 0, 0, 0, gotime.Local),
+		gotime.Date(9999, 12, 31, 0, 0, 0, 0, gotime.Local),
+	}
+	timestampRange = []gotime.Time{
+		gotime.Date(1970, 1, 1, 0, 0, 1, 0, gotime.Local),
+		gotime.Date(2038, 1, 19, 3, 14, 7, 0, gotime.Local),
```
Comment (lzmhhh123) [inline types/time.go]: Shall we consider the trailing fractional seconds referred to the docs of MySQL 8.0? It presents *`TIMESTAMP` & `DATETIME` can include a trailing fractional seconds part in up to microseconds (6 digits) precision*.

File: types/time.go (hunk 2 above)
Comment (qw4990) [reply types/time.go]: You are right.
I will fix this.

Comment (qw4990) [reply types/time.go]: Updated.

File: types/time.go (hunk 1 above)
Comment (qw4990) [reply types/time.go]: Fixed.

File: types/time.go (hunk 3 above)
Comment (qw4990) [reply types/time.go]: Yes, you are right.
Now these ranges have been removed, and it uses `MaxDatetime` and `MaxTimestamp` directly, which already take trailing fractional seconds into account.

**Comment by qw4990** [comment]
PTAL @lamxTyler

File: types/errors.go [hunk 4]
```diff
@@ -64,23 +64,24 @@ var (
...
+	codeTooBigScale              = terror.ErrCode(mysql.ErrTooBigScale)
+	codeTooBigPrecision          = terror.ErrCode(mysql.ErrTooBigPrecision)
+	codeWrongFieldSpec           = terror.ErrCode(mysql.ErrWrongFieldSpec)
+	codeTruncatedWrongValue      = terror.ErrCode(mysql.ErrTruncatedWrongValue)
+	codeUnknown                  = terror.ErrCode(mysql.ErrUnknown)
+	codeInvalidDefault           = terror.ErrCode(mysql.ErrInvalidDefault)
+	codeMBiggerThanD             = terror.ErrCode(mysql.ErrMBiggerThanD)
+	codeDataOutOfRange           = terror.ErrCode(mysql.ErrWarnDataOutOfRange)
+	codeDuplicatedValueInType    = terror.ErrCode(mysql.ErrDuplicatedValueInType)
+	codeDatetimeFunctionOverflow = terror.ErrCode(mysql.ErrDatetimeFunctionOverflow)
```
Comment (lamxTyler) [inline types/errors.go]: Where is it used?

Comment (qw4990) [reply types/errors.go]: Line 113 use it as a key in map `typesMySQLErrCodes`.

Comment (qw4990) [reply types/errors.go]: Line 113 use it as a key in map `typesMySQLErrCodes`.

Comment (lamxTyler) [reply types/errors.go]: Then can we remove it in the map since there is no one use it to query the map?

Comment (qw4990) [reply types/errors.go]: In `builtin_time.go`, line 2686 and line 2719 would return this specific error to client if it out of range.

Comment (lamxTyler) [reply types/errors.go]: No, it is not. You can see the definition of `types.ErrDatetimeFunctionOverflow`, it is defined in terms of `mysql.XXX`.

Comment (qw4990) [reply types/errors.go]: Oh, you are right, my mistake.
Now this error has been removed.

**Pull request merged**